### PR TITLE
Upgrade to Spring Data Neo4j 6

### PR DESCRIPTION
**Pull request by meistermeier**
This PR adapts Spring Boot 2.4 to Spring Data Neo4j 6.

For some background:
Neo4j has developed two projects in the last year

* A starter dedicated for the Java Driver for Neo4j (dubbed the "Bolt Driver") https://github.com/neo4j/neo4j-java-driver-spring-boot-starter
* SDN/RX: https://github.com/neo4j/sdn-rx Which does imperative and reactive Spring Data Neo4j purely based on the Bolt Driver

SDN/RX has become SDN 6 and is now merged into Spring Data Neo4j.

That means Spring Boot depending on the Ockham release train will not build anymore.

This adapts Spring Boot 2.4 to the SDN 6

* It brings automatic configuration for the Bolt Driver under the new namespace `spring.neo4j`, similar to others (like mongo and redis)
* Adapts automatic configuration for Spring Data Neo4j

This removes the dependency on Neo4j-OGM as well.

SDN 6 won't have support for bringing up an embedded instance, which is in line with our strategy to recommend Testcontainers or the Neo4j Testharness to our customers and users.

We refrained from adding support for the Neo4j Testharness to the test slices to not broaden the topic more.
However, if we want to help users to bring something up running very fast, that would be an additional option.

The change https://github.com/spring-projects/spring-boot/compare/master...meistermeier:feature/neo4j-starter?expand=1#diff-76ff4fe38f0762e0fcb54986cae8e0b6R1608 is in there to enable SDN 6 for now because the dependency to Spring Data is on Ockham M1 where SDN 6 is not available.

We are looking forward to your feedback.

### CODE REVIEW

**Comment by wilkinsona** [comment]
Thanks for the PR, @meistermeier.

This looks great but there is quite a lot to take in here. To make it easier to review, I wonder if it would be possible to split it out into separate commits or, ideally, separate pull requests please? For example, the addition of metrics looks great but I think it warrants a separate pull request.

In addition to making things easier for us to review, splitting things up will also make the changes easier to understand if we're reviewing a diff in the future. Perhaps most importantly, separate pull requests will result in a separate entries in the release notes which will hopefully make it easier for users to discover the new features.

**Comment by meistermeier** [comment]
Thanks for your quick response, Andy.
When I saw the changes in the PR, I also thought "wow" and you are right this should get splitted in sensible parts (commits or PRs).
Can wel keep this PR open for the SDN spring-boot starter update? I will go ahead and try to extract driver autoconfig and metrics into different PRs.

**Comment by wilkinsona** [comment]
> Can wel keep this PR open for the SDN spring-boot starter update? I will go ahead and try to extract driver autoconfig and metrics into different PRs.

Thanks. That sounds good to me. This PR modified so that it's just for compatibility with the latest Spring Data Ockham snapshots, and then others on top of that to introduce new features.

**Comment by meistermeier** [comment]
So I splitted this up into several PRs but unfortunately they cannot get merged independently :(
Bringing this into an order:
Neo4j Java Driver autoconfig: https://github.com/spring-projects/spring-boot/pull/22301
Spring Data Neo4j starter (this): https://github.com/spring-projects/spring-boot/pull/22299
Neo4j Actuator: https://github.com/spring-projects/spring-boot/pull/22302

Please tell me if I did something terrible wrong during the split :)

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataProperties.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	 * A statically configured database name. This property is only applicable when
+	 * connecting against a 4.0 cluster or server and will lead to errors if used with a
+	 * prior version of Neo4j. Leave this null (the default) to indicate that you like the
+	 * server to decide the default database to use. The database name set here will be
+	 * statically used throughout the lifetime of the application. If you need more
+	 * flexibility you can declare a bean of type {@link DatabaseSelectionProvider} which
+	 * can for example use Spring's Security Context or similar to determine the current
+	 * principal on which you could decide which database to use.
+	 */
+	private String database;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataProperties.java:41]: Shouldn't this be `databaseName` ?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+/**
+ * Internal configuration of Neo4j client and transaction manager.
+ *
+ * @author Michael J. Simons
+ */
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass({ Neo4jTransactionManager.class, PlatformTransactionManager.class })
+@ConditionalOnRepositoryType(store = "neo4j", type = RepositoryType.IMPERATIVE)
+@AutoConfigureAfter(Neo4jDriverAutoConfiguration.class)
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java:53]: FTR we need to merge #22301 before considering this one.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+@AutoConfigureBefore(Neo4jImperativeRepositoriesConfiguration.class)
+@Import(Neo4jDefaultCallbacksRegistrar.class)
+final class Neo4jImperativeDataConfiguration {
+
+	@Bean("databaseSelectionProvider")
+	@ConditionalOnProperty(prefix = "spring.data.neo4j", name = "database")
+	@ConditionalOnMissingBean
+	@Order(-30)
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java:61]: Can you share why this is required?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jReactiveRepositoriesConfigureRegistrar.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		return Neo4jReactiveRepositoriesConfigureRegistrar.SpringDataNeo4jRxConfiguration.class;
+	}
+
+	@Override
+	protected RepositoryConfigurationExtension getRepositoryConfigurationExtension() {
+		return new ReactiveNeo4jRepositoryConfigurationExtension();
+	}
+
+	@EnableReactiveNeo4jRepositories
+	private static class SpringDataNeo4jRxConfiguration {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jReactiveRepositoriesConfigureRegistrar.java:50]: Same here regarding the name consistency.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfiguration.java [hunk 5]
```diff
@@ -16,45 +16,29 @@
...
+ * Shared entry point for the configuration of Spring Data Neo4j repositories in their
+ * imperative and reactive forms.
  *
  * @author Dave Syer
  * @author Oliver Gierke
  * @author Josh Long
- * @since 1.4.0
+ * @author Michael J. Simons
  * @see EnableNeo4jRepositories
+ * @since 1.4.0
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfiguration.java:36]: This change doesn't seem necessary at first sight.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfiguration.java [hunk 6]
```diff
@@ -16,45 +16,29 @@
...
 @Configuration(proxyBeanMethods = false)
-@ConditionalOnClass({ Neo4jSession.class, Neo4jRepository.class })
-@ConditionalOnMissingBean({ Neo4jRepositoryFactoryBean.class, Neo4jRepositoryConfigurationExtension.class })
-@ConditionalOnProperty(prefix = "spring.data.neo4j.repositories", name = "enabled", havingValue = "true",
-		matchIfMissing = true)
-@Import(Neo4jRepositoriesRegistrar.class)
+@ConditionalOnClass(Driver.class)
 @AutoConfigureAfter(Neo4jDataAutoConfiguration.class)
+@Import({ Neo4jImperativeRepositoriesConfiguration.class, Neo4jReactiveRepositoriesConfiguration.class })
 public class Neo4jRepositoriesAutoConfiguration {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfiguration.java:42]: You could keep that type and create another auto-configuration for the reactive variant. That would be more consistent with other stores.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/MockedDriverConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import org.neo4j.driver.Driver;
+import org.neo4j.driver.Session;
+import org.neo4j.driver.SessionConfig;
+import org.neo4j.driver.types.TypeSystem;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Michael J. Simons
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/MockedDriverConfiguration.java:31]: This is missing some Javadoc

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataAutoConfigurationTests.java [hunk 8]
```diff
@@ -16,257 +16,402 @@
...
-	void defaultConfiguration() {
-		this.contextRunner.withPropertyValues("spring.data.neo4j.uri=http://localhost:8989").run((context) -> {
-			assertThat(context).hasSingleBean(org.neo4j.ogm.config.Configuration.class);
-			assertThat(context).hasSingleBean(SessionFactory.class);
-			assertThat(context).hasSingleBean(Neo4jTransactionManager.class);
-			assertThat(context).doesNotHaveBean(OpenSessionInViewInterceptor.class);
-			assertThat(context).doesNotHaveBean(BookmarkManager.class);
-		});
-	}
+	@Nested
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataAutoConfigurationTests.java:57]: We don't use `@Nested` anywhere else. I am personally not keen to introduce this style as part of this PR. I also think it makes the diff very hard to read.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfigurationIntegrationTests.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.test.context.DynamicPropertyRegistry;
+import org.springframework.test.context.DynamicPropertySource;
+import org.testcontainers.containers.Neo4jContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Michael J. Simons
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfigurationIntegrationTests.java:36]: Missing Javadoc

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfigurationTests.java [hunk 10]
```diff
@@ -44,50 +50,88 @@
...
+	void shouldRespectAtEnableReactiveNeo4jRepositories() {
+		this.contextRunner
+				.withUserConfiguration(SortOfInvalidCustomConfiguration.class, WithCustomReactiveRepositoryScan.class)
+				.withPropertyValues("spring.data.neo4j.repositories.type=reactive")
+				.run(ctx -> assertThat(ctx).doesNotHaveBean(CityRepository.class)
+						.doesNotHaveBean(ReactiveCityRepository.class).doesNotHaveBean(CountryRepository.class)
+						.hasSingleBean(ReactiveCountryRepository.class));
+	}
+
+	@Configuration
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jRepositoriesAutoConfigurationTests.java:115]: For consistency this should have `proxyBeanMethods = false`.

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/DataNeo4jTypeExcludeFilter.java [hunk 11]
```diff
@@ -31,4 +44,23 @@
...
+	static {
+		Set<Class<?>> includes = new LinkedHashSet<>();
+		includes.add(Driver.class);
+		includes.add(Neo4jClient.class);
+		includes.add(Neo4jTemplate.class);
+		includes.add(Neo4jRepository.class);
+		includes.add(ReactiveNeo4jClient.class);
+		includes.add(ReactiveNeo4jTemplate.class);
+		includes.add(ReactiveNeo4jRepository.class);
+		DEFAULT_INCLUDES = Collections.unmodifiableSet(includes);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/DataNeo4jTypeExcludeFilter.java:58]: I am a bit confused. Why is this necessary?

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+@Inherited
+@BootstrapWith(ReactiveDataNeo4jTestContextBootstrapper.class)
+@ExtendWith(SpringExtension.class)
+@OverrideAutoConfiguration(enabled = false)
+@TypeExcludeFilters(ReactiveDataNeo4jTypeExcludeFilter.class)
+@AutoConfigureCache
+@AutoConfigureDataNeo4j
+@ImportAutoConfiguration
+@Import(ReactiveCapableNeo4jVersionAssertion.class)
+public @interface ReactiveDataNeo4jTest {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java:71]: None of the other store has a reactive specific variant so I am not keen to introduce that. Perhaps the need of a specific annotation surfaces another problem?

File: spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-data-neo4j/src/main/java/smoketest/data/neo4j/CustomerRepository.java [hunk 13]
```diff
@@ -16,10 +16,10 @@
 
 package smoketest.data.neo4j;
 
-import java.util.List;
-
 import org.springframework.data.neo4j.repository.Neo4jRepository;
 
+import java.util.List;
```
Comment (snicoll) [inline spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-data-neo4j/src/main/java/smoketest/data/neo4j/CustomerRepository.java:21]: This change does not seem necessary either. I've seen a few reformat in that vein so running the build should be enough to fix them.

**Comment by meistermeier** [comment]
Thanks for the detailed feedback. My plan is to resolve the trivial ones first and see how we can align our changes more with the other Spring Data modules in the discussable cases.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataProperties.java (hunk 1 above)
Comment (meistermeier) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataProperties.java:41]: The nomenclature in the driver itself also refers to this property as `database`. So we want to align with it.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java (hunk 3 above)
Comment (meistermeier) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java:61]: It should have more importance than the `defaultSelectionProvider` underneath this method to ensure that the configured one has precedence.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeRepositoriesConfigureRegistrar.java [hunk 14]
```diff
@@ -48,7 +48,7 @@ protected RepositoryConfigurationExtension getRepositoryConfigurationExtension()
 	}
 
 	@EnableNeo4jRepositories
-	private static class EnableNeo4jRepositoriesConfiguration {
+	private static class SpringDataNeo4jRxConfiguration {
```
[the remark this answers is not in the document]
Comment (meistermeier) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeRepositoriesConfigureRegistrar.java:51]: nevermind ... ancient bits of SDN/RX naming.

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java (hunk 12 above)
Comment (meistermeier) [reply spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java:71]: I will target this together with the reactive `AutoConfiguration` suggestion in the next iteration.

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/DataNeo4jTypeExcludeFilter.java (hunk 11 above)
Comment (meistermeier) [reply spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/DataNeo4jTypeExcludeFilter.java:58]: Basically to avoid collision if a user tries to scan packages containing those classes. If I interpret your comment correctly, maybe too defensive than needed.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataProperties.java (hunk 1 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataProperties.java:36]: Fair enough. Note however that `DatabaseSelectionProvider#createStaticDatabaseSelectionProvider` has an argument called `databaseName`, not `database`. That's  what actually lead me to add this comment. The same apply for `ReactiveDatabaseSelectionProvider`.

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/DataNeo4jTypeExcludeFilter.java (hunk 11 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/DataNeo4jTypeExcludeFilter.java:58]: Yes. If you look all the other exclude type filter implementations we have, none of those do that and that's not the intention anyway.

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java (hunk 12 above)
Comment (meistermeier) [reply spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java:71]: Just checked this one and the reasons why we have chosen to introduce a parallel annotation for reactive is some convenience features:
Fail on database < 4.0 (https://github.com/spring-projects/spring-boot/pull/22299/files/b7fa3d1faca912a7c384ec3c20ca93a8fb3b21a8#diff-31539ed0ee3ec3e7c3a490aaa41e5db5R50)
and setting the `spring.data.neo4j.repositories.type` to reactive (https://github.com/spring-projects/spring-boot/pull/22299/files/b7fa3d1faca912a7c384ec3c20ca93a8fb3b21a8#diff-31539ed0ee3ec3e7c3a490aaa41e5db5R50)
Arguably this might be too less features for introducing this parallel infrastructure.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java (hunk 3 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java]: Thanks for the feedback. We don't use `@Order` that way anywhere else so I think we should express this order in a different manner. Usually we move the bits to two separate `@Configuration` that we import in order. Look, for instance, at `org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations`.

Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java]: On a second review, I think you should do this check programmatically within one single `@Bean` method, checking for the property using the `Environment`.

File: spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java (hunk 12 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/neo4j/ReactiveDataNeo4jTest.java:71]: None of the other stores have such a variant and I am not keen to introduce it as part of the upgrade. We can reconsider once the upgrade is done of course.

**Comment by snicoll** [comment]
@meistermeier @michael-simons, thank you both for helping us upgrade to Spring Data Neo4j 6! For consistency with other stores, I had to undo a number of things in your proposal, see https://github.com/snicoll/spring-boot/commit/9bc71fe44fb69f3971edcf7456cf7fc892778870 for more details.

There's a bit more work still. We need to update the documentation (#22652) and I'd like to experiment supporting the `spring.data.neo4j` properties again (i.e. the ones used to bootstrap the connection with the client) (#22653)?

**Comment by jagadeesh595** [comment]
With the latest version of SDN 6.0.0-M2, I have below observation as of now. When we go with either dynamic relationship or Relationship Properties, records are saving in database quickly, but retrieval taking more time compared to Old SDN 5.3.3 version.

In new version, below are the syntaxes to create either dynamic relationship or Relationship Properties,

Syntax for relationship with properties: Below syntax represents Sponsor relationship between two member nodes. MemberSponsor entity contains all relationship properties.

@Relationship(value = "sponsor", direction = Relationship.Direction.OUTGOING)
private Map<Member, MemberSponsor> hasSponsor = new HashMap<>();
Syntax for dynamic relationship without any properties: Below syntax represents Sponsor relationship between two member nodes. In the Map, key-String represents the relationship name, that we can configure dynamically.

@Relationship(direction = Relationship.Direction.OUTGOING)
private Map<String, Member> sponsors = new HashMap<>();
When I am trying to get Member data using below JPA method, system taking more time when any of the above relationships exists in the Member entity class. It is taking 25 mins to get the single Member data using below method from latest version SDN 6.0.0. In old version SDN 5.3.3 – getting same member data within a second. This issue especially occurring only when there is a relationship exists between two nodes with same labels. Here, we are referring relationship between two member nodes of same label.

import org.springframework.data.neo4j.repository.Neo4jRepository;

public interface MemberRepository extends Neo4jRepository<Member, Long> {
    Member findByMemberId(String memberId);
}
Can you please help me on this. Seems like issue with retrieval when there is a Dynamic relationship on entity.

**Comment by meistermeier** [comment]
Thanks @jagadeesh595 for reporting this.
Such problems are better placed in the official issue tracker of Spring Data Neo4j: https://jira.spring.io/projects/DATAGRAPH/
I created a ticket for this problem: https://jira.spring.io/browse/DATAGRAPH-1372

**Comment by jagadeesh595** [comment]
> Thanks @jagadeesh595 for reporting this.
> Such problems are better placed in the official issue tracker of Spring Data Neo4j: https://jira.spring.io/projects/DATAGRAPH/
> I created a ticket for this problem: https://jira.spring.io/browse/DATAGRAPH-1372

Thank you for the quick update.

**Comment by jagadeesh595** [comment]
I am not seeing any response for above question. Hope, we need to reach out to neo4j engineering team.

**Comment by snicoll** [comment]
@jagadeesh595 there is nothing we can do here I am afraid. Please comment on the Jira issue that is referenced above.

**Comment by michael-simons** [comment]
Sorry, @snicoll but I like to address this as well.

@jagadeesh595 Thanks for your interested in the new version of SDN coming up. We appreciate that a lot. Please be aware that the Neo4j engineering team is already reading this thread. Also we understand that you are in contact with one of our colleagues from sale. We are answering those questions as well. And we had an eye on the Jira ticket. @meistermeier already sent an E-Mail out to answer all those questions. So, as much as we appreciate your great analysis, let us keep the discussion either indirectly via email or directly on the raised Jira issue. Thanks so much.

To sum it up: When in doubt, use the SDN+OGM coming with Boot 2.3 **Our recommendations for new projects is however SDN 6. All the issues raised will be addressed.**

**Comment by jagadeesh595** [comment]
> Sorry, @snicoll but I like to address this as well.
> 
> @jagadeesh595 Thanks for your interested in the new version of SDN coming up. We appreciate that a lot. Please be aware that the Neo4j engineering team is already reading this thread. Also we understand that you are in contact with one of our colleagues from sale. We are answering those questions as well. And we had an eye on the Jira ticket. @meistermeier already sent an E-Mail out to answer all those questions. So, as much as we appreciate your great analysis, let us keep the discussion either indirectly via email or directly on the raised Jira issue. Thanks so much.
> 
> To sum it up: When in doubt, use the SDN+OGM coming with Boot 2.3 **Our recommendations for new projects is however SDN 6. All the issues raised will be addressed.**

Thanks you Michael! We received answers from your team and communicating through email.
